### PR TITLE
Automatically clean up soft-deleted records after X days.

### LINKED ISSUE #14726: Automatically empty the Trash
@charlesBochet told me we were moving to applying unicity contraints to all objects, including soft-deleted objects.

That might be annoying, so I think that in that case, we should automatically empty the trash (default value = 14 days). Salesforce does it every 15 days.

We should let the workspace owner change this default value (0 to xxx value, no reason to set a limit)
Even though it doesn't git well, I think we should hide it this in "Security" for now.

cc @Bonapara

### PR DESCRIPTION
**Pull request by mabdullahabaid**
Closes #14726

### Added
- `trashRetentionDays` field to workspace entity (default: 14 days)  
- Automated trash cleanup using BullMQ jobs  
- Daily cron (00:10 UTC) that enqueues cleanup jobs for all active workspaces  
- Per-workspace limit: 100k records deleted per day  
- Calendar-based retention: records deleted on day X are cleaned up X+14 days later (at midnight UTC boundaries)  

### Architecture
- **Cron (WorkspaceTrashCleanupCronJob):** Runs daily, enqueues jobs in parallel for all workspaces  
- **Job (WorkspaceTrashCleanupJob):** Processes individual workspace cleanup  
- **Service (WorkspaceTrashCleanupService):** Discovers tables with `deletedAt`, deletes old records with quota enforcement  
- **Command:** `npx nx run twenty-server:command cron:workspace:cleanup-trash` to register the cron  

### Testing
- Unit tests for service with 100% coverage of public API  
- Tested quota enforcement, error handling, and edge cases

### CODE REVIEW

File: packages/twenty-server/src/engine/trash-cleanup/services/trash-cleanup.service.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    return totalDeleted;
+  }
+
+  private calculateCutoffDate(trashRetentionDays: number): Date {
+    const cutoffDate = new Date();
+
+    cutoffDate.setUTCHours(0, 0, 0, 0);
+    cutoffDate.setDate(cutoffDate.getDate() - trashRetentionDays + 1);
```
[the remark this answers is not in the document]
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/trash-cleanup/services/trash-cleanup.service.ts:142]: That's intentional. If the CRON runs on October 15, 2025, we want to delete all records with deletedAt `< October 2, 2025` so that records deleted on October 1, 2025 are hard-deleted.

File: packages/twenty-front/src/modules/users/graphql/fragments/userQueryFragment.ts [hunk 2]
```diff
@@ -79,6 +79,10 @@ export const USER_QUERY_FRAGMENT = gql`
         id
       }
       isTwoFactorAuthenticationEnforced
+      trashRetentionDays
+      views {
+        ...ViewFragment
```
[the remark this answers is not in the document]
Comment (mabdullahabaid) [reply packages/twenty-front/src/modules/users/graphql/fragments/userQueryFragment.ts]: Not sure what this means and if it has anything to do with my changes. If it does, please let me know what this really means as in why is it needed..

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/workspace-trash-cleanup.module.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+import { Workspace } from 'src/engine/core-modules/workspace/workspace.entity';
+import { ObjectMetadataEntity } from 'src/engine/metadata-modules/object-metadata/object-metadata.entity';
+import { WorkspaceTrashCleanupCronCommand } from 'src/engine/workspace-manager/workspace-trash-cleanup/commands/workspace-trash-cleanup.cron.command';
+import { WorkspaceTrashCleanupCronJob } from 'src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job';
+import { WorkspaceTrashCleanupJob } from 'src/engine/workspace-manager/workspace-trash-cleanup/jobs/workspace-trash-cleanup.job';
+import { WorkspaceTrashCleanupService } from 'src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service';
+
+@Module({
+  imports: [TypeOrmModule.forFeature([Workspace, ObjectMetadataEntity])],
```
[the remark this answers is not in the document]
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/workspace-trash-cleanup.module.ts]: Added.

File: packages/twenty-front/src/pages/settings/security/SettingsSecurity.tsx [hunk 4]
```diff
@@ -32,8 +36,50 @@ const StyledSection = styled(Section)`
...
+
+      setCurrentWorkspace({
+        ...currentWorkspace,
+        trashRetentionDays: numValue,
+      });
+    } catch (err: any) {
+      enqueueErrorSnackBar({
+        apolloError: err instanceof ApolloError ? err : undefined,
+      });
+    }
```
[the remark this answers is not in the document]
Comment (mabdullahabaid) [reply packages/twenty-front/src/pages/settings/security/SettingsSecurity.tsx]: Done.

**Comment by mabdullahabaid** [comment]
I feel my config is properly set and this has to do with the CI - run already passed for [5fff113](https://github.com/twentyhq/twenty/pull/14862/commits/5fff1132b2ba02fa8425e3c70798ee22b16598cd).

<p align="center">
image
</p>

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    );
+  }
+
+  async cleanupWorkspaceTrash(
+    input: WorkspaceTrashCleanupInput,
+  ): Promise<WorkspaceTrashCleanupResult> {
+    const { workspaceId, schemaName, trashRetentionDays } = input;
+
+    try {
+      const tableNames = await this.discoverTablesWithSoftDelete(schemaName);
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service.ts]: I don't think we gain anything by first FETCHING before DELETING? My intuition is that it's just more costly to do it this way but maybe I'm missing something?

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+      );
+
+      return { success: false, deletedCount: 0, error: errorMessage };
+    }
+  }
+
+  private async discoverTablesWithSoftDelete(
+    schemaName: string,
+  ): Promise<string[]> {
+    const rawResults = await this.objectMetadataRepository
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service.ts]: Maybe it makes sens to use the raw query builder here but if you can use high-level methods that's always better. We might not keep this "find tables" piece anyway

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service.ts [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    return cutoffDate;
+  }
+
+  private async deleteFromTable(
+    schemaName: string,
+    tableName: string,
+    cutoffDate: Date,
+    limit: number,
+  ): Promise<number> {
+    const result = await this.dataSource.query(
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service.ts]: Typically an example where you're going much too low in abstraction. You're loosing a lot of value by doing this.

For example in `WorkspaceDeleteQueryBuilder` which is called  when you call `delete()`, you'll see there's an `eventEmitterService`. By going "raw sql" you lose this so people won't get destroy webhooks / there will be no destroy events (maybe there's some cascade deleting happening or other parts of the app that want to be aware you destroyed records)

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      workspaces.map((workspace) =>
+        this.messageQueueService.add<WorkspaceTrashCleanupJobData>(
+          WorkspaceTrashCleanupJob.name,
+          {
+            workspaceId: workspace.id,
+            schemaName: workspace.schema,
+            trashRetentionDays: workspace.trashRetentionDays,
+          },
+          {
+            priority: 10,
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts]: Do we often pass priority like that? What made you pick 10?

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+      .createQueryBuilder('workspace')
+      .innerJoin(
+        DataSourceEntity,
+        'dataSource',
+        'dataSource.workspaceId = workspace.id',
+      )
+      .where('workspace.activationStatus = :status', {
+        status: WorkspaceActivationStatus.ACTIVE,
+      })
+      .andWhere('workspace.trashRetentionDays >= :minDays', { minDays: 0 })
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts]: What does it mean to have a negative trashRetentionDays / why are we checking for that?

File: packages/twenty-front/src/modules/settings/components/SettingsOptions/__stories__/SettingsOptionCardContentInput.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+const StyledContainer = styled.div`
+  width: 480px;
+`;
+
+const SettingsOptionCardContentInputWrapper = (
+  args: React.ComponentProps<typeof SettingsOptionCardContentInput>,
+) => {
+  const handleBlur = (value: string) => {
+    // eslint-disable-next-line no-console
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/settings/components/SettingsOptions/__stories__/SettingsOptionCardContentInput.stories.tsx]: Leftover?

File: packages/twenty-front/src/modules/settings/components/SettingsOptions/SettingsOptionCardContentInput.tsx [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  flex-direction: column;
+`;
+
+type SettingsOptionCardContentInputProps = {
+  Icon?: IconComponent;
+  title: React.ReactNode;
+  description?: string;
+  disabled?: boolean;
+  value: string;
+  onBlur: (value: string) => void;
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/settings/components/SettingsOptions/SettingsOptionCardContentInput.tsx]: Is this onblur a common pattern in the codebase? I don't remember seeing it before

File: packages/twenty-front/src/modules/settings/components/SettingsOptions/SettingsOptionCardContentInput.tsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  placeholder?: string;
+};
+
+export const SettingsOptionCardContentInput = ({
```
Comment (FelixMalfait) [inline packages/twenty-front/src/modules/settings/components/SettingsOptions/SettingsOptionCardContentInput.tsx]: I see we had a similar SettingsOptionCardContentCounter component, why not use it? I did not checkout the app and look into it so there might be good reasons but I prefer asking. Feel free to challenge things with @Bonapara he will always appreciate when there is an opportunity to avoid introducing a new component/pattern

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+    this.logger.log(
+      `Successfully enqueued ${workspaces.length} trash cleanup job(s)`,
+    );
+  }
+
+  private async getActiveWorkspaces(): Promise<
+    Array<{ id: string; trashRetentionDays: number; schema: string }>
+  > {
+    const rawResults = await this.workspaceRepository
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts]: Prefer non-raw if possible (maybe not!)

Comment (FelixMalfait) [reply packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts]: You should be proper typing etc

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/jobs/workspace-trash-cleanup.job.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
+import { Injectable, Logger } from '@nestjs/common';
+
+import { Process } from 'src/engine/core-modules/message-queue/decorators/process.decorator';
+import { Processor } from 'src/engine/core-modules/message-queue/decorators/processor.decorator';
+import { MessageQueue } from 'src/engine/core-modules/message-queue/message-queue.constants';
+import { WorkspaceTrashCleanupService } from 'src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service';
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/jobs/workspace-trash-cleanup.job.ts]: I think I would not consider this a workspace-manager service, simply call it trash-cleanup and put it one level above? Workspace Manager tend to be a complex area so maybe best not too overload it with simpler stuff like this for now. But honestly it wasn't a bad choice, you could argue both

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts (hunk 8 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts]: This made the CRON low-priority, but I see we're not passing priority anywhere else, allowing the default FIFO order.

This is over-engineering and over-optimization on my part for an operation that already has a limit applied (100k records) and runs once per day. 

Apologies.

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts (hunk 9 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/crons/workspace-trash-cleanup.cron.job.ts]: This is also an over-optimization and was added just in case we want to allow -1 to disable trash cleanup in the future. 

However, I also added a minimum to the database field so that we cannot store negative values for trashRetentionDays.

From a product perspective, I think we may never allow trashRetention to be turned off, so this is redundant. 

Fixing.

File: packages/twenty-front/src/modules/settings/components/SettingsOptions/__stories__/SettingsOptionCardContentInput.stories.tsx (hunk 10 above)
Comment (mabdullahabaid) [reply packages/twenty-front/src/modules/settings/components/SettingsOptions/__stories__/SettingsOptionCardContentInput.stories.tsx]: It was not a leftover, but I see other stories test callback functions differently. Will fix if we decide to keep a separate component for Input, otherwise the story itself will become redundant anyway.

File: packages/twenty-front/src/modules/settings/components/SettingsOptions/SettingsOptionCardContentInput.tsx (hunk 11 above)
Comment (mabdullahabaid) [reply packages/twenty-front/src/modules/settings/components/SettingsOptions/SettingsOptionCardContentInput.tsx]: Nope. I added it to reduce network calls due to keystrokes, but upon rechecking, I believe you prefer de-bouncing instead of onBlur. I will update this too then to follow the pattern.

File: packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/jobs/workspace-trash-cleanup.job.ts (hunk 14 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/jobs/workspace-trash-cleanup.job.ts]: I added it to workspace-manager since it was performing admin work on the workspace. 

Seemed logical, but it's also fair to move it up since workspace-manager handles the complex bit.

File: packages/twenty-front/src/modules/settings/components/SettingsOptions/SettingsOptionCardContentInput.tsx (hunk 12 above)
Comment (mabdullahabaid) [reply packages/twenty-front/src/modules/settings/components/SettingsOptions/SettingsOptionCardContentInput.tsx]: The component is a little different.

image

This is what we need.

494006347-67b39117-5d49-4fd8-8a3b-c1cf3a444f31

SettingsOptionCardContentCounter can be updated by making the minus and the plus optional via props. One more thing that appears different is the alignment of number inside the input field, but I think that's a small discrepancy between code and design.

I agree, I should have re-used the component by making some edits. 

Another rookie mistake that I need to avoid going forward.

**Comment by mabdullahabaid** [comment]
I think given the number of mistakes I made in this issue, I need to re-think a few things from SQL queries to re-usability to types.

Will push an update soon and try to trace things better.

**Comment by mabdullahabaid** [comment]
Resolved a few comments by closely following how other cron.job.ts and job.ts files are set. 

Also replaced the SettingsOptionCardContentInput with SettingsOptionCardContentCounter and updated story.

Still working on the service, will leave a comment once done.

**Comment by mabdullahabaid** [comment]
Ready for another review. 

Re-thought the solution to go through TwentyORM. It's using cached version of database table names to reduce the round-trip, but cleaning up in configurable batches (1000 as default) to avoid database locks with massive queries.

- The reason we rely on discovering database tables is to access their repositories.
- We need access to repositories because we need to find IDs - TypeORM delete would not allow limits and orderBy.

Find is an extra query here. Say that we have 50 tables and only the last one has 150k deletedAt records < cutoff date.

We go through 49 queries for find and short-circuit when we find nothing to delete, then another 200 to delete 100k records in chunks of 1000 (until we increase the chunk size). So, a total of 249 queries here per workspace in the worst-case. Increase objects to 100 and the total goes to 299 if we have over 100k records to clean-up in the final table.

Scaling this seems reasonable, but open to comments.

One more case that I have not yet covered in this PR: when trashRetention is 0, softDelete should not come into the picture. As per current implementation, even at 0, the next CRON clears the record out. I think we can cover this in a separate PR, unless you want otherwise.

For webhooks, even though I started going through the TwentyORM layer, I was not able to receive anything at https://webhook.site when the CRON ran. I tried selecting `delete` and `all` on the UI. 

Will appreciate some help/clarity on the webhook thing - do we have a destroy event for hard-delete and if we do, should it not have run when I set to "all"?

File: packages/twenty-server/src/engine/core-modules/workspace/dtos/update-workspace-input.ts [hunk 15]
```diff
@@ -197,4 +199,10 @@ export class UpdateWorkspaceInput {
   @IsBoolean()
   @IsOptional()
   isTwoFactorAuthenticationEnforced?: boolean;
+
+  @Field({ nullable: true })
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/core-modules/workspace/dtos/update-workspace-input.ts:203]: Not nullable in the UI or in the DB I think? Are you sure it should be nullable here?

File: packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts [hunk 16]
```diff
@@ -1129,6 +1129,25 @@ export class ConfigVariables {
...
 
+  @ConfigVariablesMetadata({
+    group: ConfigVariablesGroup.Other,
+    description:
+      'Maximum number of records to delete per workspace during trash cleanup',
+    type: ConfigVariableType.NUMBER,
+  })
+  @CastToPositiveNumber()
+  @IsOptional()
+  TRASH_CLEANUP_MAX_RECORDS_PER_WORKSPACE = 100000;
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts]: I don't think this deserves to be a config var as I don't see us changing that every day. The cron pattern would be more likely to change than this imo. It's not a blocker for merge and subjective but just my opinion :)

File: packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts [hunk 17]
```diff
@@ -1129,6 +1129,25 @@ export class ConfigVariables {
...
+
+  @ConfigVariablesMetadata({
+    group: ConfigVariablesGroup.Other,
+    description: 'Number of records deleted per batch during trash cleanup',
+    type: ConfigVariableType.NUMBER,
+  })
+  @CastToPositiveNumber()
+  @IsOptional()
+  TRASH_CLEANUP_BATCH_SIZE = 1000;
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts]: Same comment as above

File: packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts (hunk 16 above)
Comment (FelixMalfait) [reply packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts]: (don't add the cron pattern if you change, I think even the cron pattern doesn't really need to be changed as config)

Comment (FelixMalfait) [reply packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts]: We can set the value to 1_000_000

File: packages/twenty-server/src/engine/trash-cleanup/crons/trash-cleanup.cron.job.ts [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        );
+      } catch (error) {
+        this.exceptionHandlerService.captureExceptions([error], {
+          workspace: {
+            id: workspace.id,
+          },
+        });
+      }
+    }
+
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/trash-cleanup/crons/trash-cleanup.cron.job.ts:66]: In Harshit's open PR we set priorities for queues. I think this will help a lot for this kind of use-case to avoid clogging. Typically here we want to set a very low priority

File: packages/twenty-server/src/database/typeorm/core/migrations/common/1759182953990-add-workspace-trash-retention.ts [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+  implements MigrationInterface
+{
+  name = 'AddWorkspaceTrashRetention1759182953990';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `ALTER TABLE "core"."workspace" ADD "trashRetentionDays" integer NOT NULL DEFAULT '14'`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."workspace" ADD CONSTRAINT "trash_retention_positive" CHECK ("trashRetentionDays" >= 0)`,
```
Comment (FelixMalfait) [inline packages/twenty-server/src/database/typeorm/core/migrations/common/1759182953990-add-workspace-trash-retention.ts]: It's debatable but I wouldn't add this, it's not important enough. It will slow down writes for something that is anecdotal (there could be 100 parameters like this one in 3 years on the workspace table). We do love to add constraints but it needs to be a critical path (think <5 contraints per table long-term)

File: packages/twenty-server/src/engine/core-modules/workspace/dtos/update-workspace-input.ts (hunk 15 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/core-modules/workspace/dtos/update-workspace-input.ts:203]: I set it nullable for the UI to be able to call the mutation without a value for trashRetentionDays, but I will revisit the logic to confirm.

File: packages/twenty-server/src/engine/trash-cleanup/crons/trash-cleanup.cron.job.ts (hunk 18 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/trash-cleanup/crons/trash-cleanup.cron.job.ts:66]: He has defined a file called `message-queue-priority.constants.ts` to hold priority values. I believe his PR would need to be merged for me to assign a priority in the same file.

File: packages/twenty-server/src/database/typeorm/core/migrations/common/1759182953990-add-workspace-trash-retention.ts (hunk 19 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/database/typeorm/core/migrations/common/1759182953990-add-workspace-trash-retention.ts]: Makes sense. DTO validation should be enough, we can avoid the database-level constraint.

File: packages/twenty-server/src/engine/core-modules/workspace/dtos/update-workspace-input.ts (hunk 15 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/core-modules/workspace/dtos/update-workspace-input.ts:203]: I looked into this and I found that we have set nullable for a bunch of other fields in the input DTO.

image

<br />

image

It is such even though we have default values for those fields. 

image

Sending null intentionally from the mutation throws a database error for all of them (unless null is allowed), but if we omit those fields in the mutation, database does not cause any issues since we're not updating the value for those fields.

image

<br />

image

<br />

image

We can handle it gracefully by removing fields containing "null" at runtime, but it's going to be a larger refactor than just trashRetentionLogic - instead of adding if-statements inside the service method, the ideal solution would be to use some utility function and then test other DTOs as well.

That said, trashRetentionDays follows the existing pattern and we are not inserting null into the database, we're just allowing the mutation to "not send" trashRetentionDays when the mutation is called.

If I remove { nullable: true } from the Field decorator, I get the following when updating the displayName.

image

File: packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts (hunk 16 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts]: Done.

File: packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts (hunk 17 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/engine/core-modules/twenty-config/config-variables.ts]: Done.

File: packages/twenty-server/src/database/typeorm/core/migrations/common/1759182953990-add-workspace-trash-retention.ts (hunk 19 above)
Comment (mabdullahabaid) [reply packages/twenty-server/src/database/typeorm/core/migrations/common/1759182953990-add-workspace-trash-retention.ts]: Done.

File: packages/twenty-server/src/engine/core-modules/workspace/dtos/update-workspace-input.ts (hunk 15 above)
Comment (FelixMalfait) [reply packages/twenty-server/src/engine/core-modules/workspace/dtos/update-workspace-input.ts:203]: Oh okay, thanks for these details!

File: packages/twenty-server/src/engine/trash-cleanup/constants/trash-cleanup.constants.ts [hunk 20]
```diff
@@ -1,2 +1,5 @@
 // Daily at 00:10 UTC
 export const TRASH_CLEANUP_CRON_PATTERN = '10 0 * * *';
+
+export const TRASH_CLEANUP_MAX_RECORDS_PER_WORKSPACE = 1_000_000;
```
Comment (FelixMalfait) [inline packages/twenty-server/src/engine/trash-cleanup/constants/trash-cleanup.constants.ts:4]: We try to keep one export per file. I will merge this PR to move on but could you please include this fix in any of your future PRs?

**Review by FelixMalfait — APPROVED** [review]
